pkg/route/ingress/ingress.go

```diff
@@ -565,6 +572,9 @@ func newRouteForIngress(
 
 func preserveRouteAttributesFromExisting(r, existing *routev1.Route) {
 	r.Name = existing.Name
+	if r.Annotations != nil && existing.Annotations != nil {
```

Comment:
I think `if r.Annotations != nil && existing.Annotations != nil` should be deleted for the same reason I mentioned on the patch logic.

Reply:
The unit test cases fails after making these changes ->

[miheer@miheer openshift-controller-manager]$ GO111MODULE=on go test -v -run TestController_sync ./pkg/route/ingress/
=== RUN   TestController_sync
=== RUN   TestController_sync/no_changes
=== RUN   TestController_sync/sync_namespace_-_no_ingress
=== RUN   TestController_sync/sync_namespace_-_two_ingress
=== RUN   TestController_sync/ignores_incomplete_ingress_-_no_host
=== RUN   TestController_sync/ignores_incomplete_ingress_-_no_service
=== RUN   TestController_sync/ignores_incomplete_ingress_-_no_paths
=== RUN   TestController_sync/ignores_incomplete_ingress_-_service_does_not_exist
=== RUN   TestController_sync/create_route
=== RUN   TestController_sync/create_route_-_targetPort_string,_service_port_with_name
=== RUN   TestController_sync/create_route_-_blocked_by_expectation
=== RUN   TestController_sync/update_route
=== RUN   TestController_sync/no-op
=== RUN   TestController_sync/no-op_-_ignore_partially_owned_resource
=== RUN   TestController_sync/update_ingress_with_missing_secret_ref
=== RUN   TestController_sync/update_ingress_to_not_reference_secret
=== RUN   TestController_sync/update_route_-_tls_config_missing
=== RUN   TestController_sync/update_route_-_termination_policy_changed_to_passthrough
=== RUN   TestController_sync/update_route_-_termination_policy_changed_to_reencrypt
=== RUN   TestController_sync/update_route_-_termination_policy_changed_to_reencrypt_and_no_tls_secret
=== RUN   TestController_sync/termination_policy_on_ingress_invalid,_nothing_happens
=== RUN   TestController_sync/termination_policy_on_ingress_invalid,_disables_tls
=== RUN   TestController_sync/Empty_tlsconfig_enables_edge_termination_without_explicit_cert
=== RUN   TestController_sync/update_route_-_secret_values_changed
=== RUN   TestController_sync/no-op_-_has_TLS
=== RUN   TestController_sync/no-op_-_has_secret_with_empty_keys
=== RUN   TestController_sync/no-op_-_termination_policy_has_been_changed_by_the_user
=== RUN   TestController_sync/update_route_-_router_admitted_route
=== RUN   TestController_sync/update_route_-_second_router_admitted_route
=== RUN   TestController_sync/no-op_-_ingress_status_already_updated
=== RUN   TestController_sync/no-op_-_router_rejected_route
=== RUN   TestController_sync/delete_route_when_referenced_secret_is_not_TLS
=== RUN   TestController_sync/delete_route_when_referenced_secret_is_not_valid
=== RUN   TestController_sync/ignore_route_when_parent_ingress_no_longer_exists_(gc_will_handle)
=== RUN   TestController_sync/update_route_-_termination_policy_changed_to_passthrough_and_timeout_set
--- FAIL: TestController_sync (0.00s)
    --- PASS: TestController_sync/no_changes (0.00s)
    --- PASS: TestController_sync/sync_namespace_-_no_ingress (0.00s)
    --- PASS: TestController_sync/sync_namespace_-_two_ingress (0.00s)
    --- PASS: TestController_sync/ignores_incomplete_ingress_-_no_host (0.00s)
    --- PASS: TestController_sync/ignores_incomplete_ingress_-_no_service (0.00s)
    --- PASS: TestController_sync/ignores_incomplete_ingress_-_no_paths (0.00s)
    --- PASS: TestController_sync/ignores_incomplete_ingress_-_service_does_not_exist (0.00s)
    --- PASS: TestController_sync/create_route (0.00s)
    --- PASS: TestController_sync/create_route_-_targetPort_string,_service_port_with_name (0.00s)
    --- PASS: TestController_sync/create_route_-_blocked_by_expectation (0.00s)
    --- PASS: TestController_sync/update_route (0.00s)
    --- PASS: TestController_sync/no-op (0.00s)
    --- PASS: TestController_sync/no-op_-_ignore_partially_owned_resource (0.00s)
    --- PASS: TestController_sync/update_ingress_with_missing_secret_ref (0.00s)
    --- PASS: TestController_sync/update_ingress_to_not_reference_secret (0.00s)
    --- PASS: TestController_sync/update_route_-_tls_config_missing (0.00s)
    --- FAIL: TestController_sync/update_route_-_termination_policy_changed_to_passthrough (0.00s)
        ingress_test.go:2603: unexpected action[0]: [{"op":"replace","path":"/spec","value":{"host":"test.com","path":"/","to":{"kind":"","name":"service-1","weight":null},"port":{"targetPort":"http"},"tls":{"termination":"passthrough","insecureEdgeTerminationPolicy":"Redirect"}}},{"op":"replace","path":"/metadata/annotations","value":null}]
    --- FAIL: TestController_sync/update_route_-_termination_policy_changed_to_reencrypt (0.00s)
        ingress_test.go:2603: unexpected action[0]: [{"op":"replace","path":"/spec","value":{"host":"test.com","path":"/","to":{"kind":"","name":"service-1","weight":null},"port":{"targetPort":"http"},"tls":{"termination":"reencrypt","certificate":"cert","key":"key","insecureEdgeTerminationPolicy":"Redirect"}}},{"op":"replace","path":"/metadata/annotations","value":null}]
    --- FAIL: TestController_sync/update_route_-_termination_policy_changed_to_reencrypt_and_no_tls_secret (0.00s)
        ingress_test.go:2603: unexpected action[0]: [{"op":"replace","path":"/spec","value":{"host":"test.com","path":"/","to":{"kind":"","name":"service-1","weight":null},"port":{"targetPort":"http"},"tls":{"termination":"reencrypt","insecureEdgeTerminationPolicy":"Redirect"}}},{"op":"replace","path":"/metadata/annotations","value":null}]
    --- FAIL: TestController_sync/termination_policy_on_ingress_invalid,_nothing_happens (0.00s)
        ingress_test.go:2623: Controller.sync() unexpected actions: []testing.Action{testing.PatchActionImpl{ActionImpl:testing.ActionImpl{Namespace:"test", Verb:"patch", Resource:schema.GroupVersionResource{Group:"route.openshift.io", Version:"v1", Resource:"routes"}, Subresource:""}, Name:"1-abcdef", PatchType:"application/json-patch+json", Patch:[]uint8{0x5b, 0x7b, 0x22, 0x6f, 0x70, 0x22, 0x3a, 0x22, 0x72, 0x65, 0x70, 0x6c, 0x61, 0x63, 0x65, 0x22, 0x2c, 0x22, 0x70, 0x61, 0x74, 0x68, 0x22, 0x3a, 0x22, 0x2f, 0x73, 0x70, 0x65, 0x63, 0x22, 0x2c, 0x22, 0x76, 0x61, 0x6c, 0x75, 0x65, 0x22, 0x3a, 0x7b, 0x22, 0x68, 0x6f, 0x73, 0x74, 0x22, 0x3a, 0x22, 0x74, 0x65, 0x73, 0x74, 0x2e, 0x63, 0x6f, 0x6d, 0x22, 0x2c, 0x22, 0x70, 0x61, 0x74, 0x68, 0x22, 0x3a, 0x22, 0x2f, 0x22, 0x2c, 0x22, 0x74, 0x6f, 0x22, 0x3a, 0x7b, 0x22, 0x6b, 0x69, 0x6e, 0x64, 0x22, 0x3a, 0x22, 0x22, 0x2c, 0x22, 0x6e, 0x61, 0x6d, 0x65, 0x22, 0x3a, 0x22, 0x73, 0x65, 0x72, 0x76, 0x69, 0x63, 0x65, 0x2d, 0x31, 0x22, 0x2c, 0x22, 0x77, 0x65, 0x69, 0x67, 0x68, 0x74, 0x22, 0x3a, 0x6e, 0x75, 0x6c, 0x6c, 0x7d, 0x2c, 0x22, 0x70, 0x6f, 0x72, 0x74, 0x22, 0x3a, 0x7b, 0x22, 0x74, 0x61, 0x72, 0x67, 0x65, 0x74, 0x50, 0x6f, 0x72, 0x74, 0x22, 0x3a, 0x22, 0x68, 0x74, 0x74, 0x70, 0x22, 0x7d, 0x2c, 0x22, 0x74, 0x6c, 0x73, 0x22, 0x3a, 0x7b, 0x22, 0x74, 0x65, 0x72, 0x6d, 0x69, 0x6e, 0x61, 0x74, 0x69, 0x6f, 0x6e, 0x22, 0x3a, 0x22, 0x65, 0x64, 0x67, 0x65, 0x22, 0x2c, 0x22, 0x63, 0x65, 0x72, 0x74, 0x69, 0x66, 0x69, 0x63, 0x61, 0x74, 0x65, 0x22, 0x3a, 0x22, 0x63, 0x65, 0x72, 0x74, 0x22, 0x2c, 0x22, 0x6b, 0x65, 0x79, 0x22, 0x3a, 0x22, 0x6b, 0x65, 0x79, 0x22, 0x2c, 0x22, 0x69, 0x6e, 0x73, 0x65, 0x63, 0x75, 0x72, 0x65, 0x45, 0x64, 0x67, 0x65, 0x54, 0x65, 0x72, 0x6d, 0x69, 0x6e, 0x61, 0x74, 0x69, 0x6f, 0x6e, 0x50, 0x6f, 0x6c, 0x69, 0x63, 0x79, 0x22, 0x3a, 0x22, 0x52, 0x65, 0x64, 0x69, 0x72, 0x65, 0x63, 0x74, 0x22, 0x7d, 0x7d, 0x7d, 0x2c, 0x7b, 0x22, 0x6f, 0x70, 0x22, 0x3a, 0x22, 0x72, 0x65, 0x70, 0x6c, 0x61, 0x63, 0x65, 0x22, 0x2c, 0x22, 0x70, 0x61, 0x74, 0x68, 0x22, 0x3a, 0x22, 0x2f, 0x6d, 0x65, 0x74, 0x61, 0x64, 0x61, 0x74, 0x61, 0x2f, 0x61, 0x6e, 0x6e, 0x6f, 0x74, 0x61, 0x74, 0x69, 0x6f, 0x6e, 0x73, 0x22, 0x2c, 0x22, 0x76, 0x61, 0x6c, 0x75, 0x65, 0x22, 0x3a, 0x6e, 0x75, 0x6c, 0x6c, 0x7d, 0x5d}}}
    --- FAIL: TestController_sync/termination_policy_on_ingress_invalid,_disables_tls (0.00s)
        ingress_test.go:2603: unexpected action[0]: [{"op":"replace","path":"/spec","value":{"host":"test.com","path":"/","to":{"kind":"","name":"service-1","weight":null},"port":{"targetPort":"http"}}},{"op":"replace","path":"/metadata/annotations","value":null}]
    --- PASS: TestController_sync/Empty_tlsconfig_enables_edge_termination_without_explicit_cert (0.00s)
    --- PASS: TestController_sync/update_route_-_secret_values_changed (0.00s)
    --- PASS: TestController_sync/no-op_-_has_TLS (0.00s)
    --- PASS: TestController_sync/no-op_-_has_secret_with_empty_keys (0.00s)
    --- PASS: TestController_sync/no-op_-_termination_policy_has_been_changed_by_the_user (0.00s)
    --- PASS: TestController_sync/update_route_-_router_admitted_route (0.00s)
    --- PASS: TestController_sync/update_route_-_second_router_admitted_route (0.00s)
    --- PASS: TestController_sync/no-op_-_ingress_status_already_updated (0.00s)
    --- PASS: TestController_sync/no-op_-_router_rejected_route (0.00s)
    --- PASS: TestController_sync/delete_route_when_referenced_secret_is_not_TLS (0.00s)
    --- PASS: TestController_sync/delete_route_when_referenced_secret_is_not_valid (0.00s)
    --- PASS: TestController_sync/ignore_route_when_parent_ingress_no_longer_exists_(gc_will_handle) (0.00s)
    --- FAIL: TestController_sync/update_route_-_termination_policy_changed_to_passthrough_and_timeout_set (0.00s)
        ingress_test.go:2603: unexpected action[0]: [{"op":"replace","path":"/spec","value":{"host":"test.com","path":"/","to":{"kind":"","name":"service-1","weight":null},"port":{"targetPort":"http"},"tls":{"termination":"passthrough","insecureEdgeTerminationPolicy":"Redirect"}}},{"op":"replace","path":"/metadata/annotations","value":null}]
FAIL
FAIL	github.com/openshift/openshift-controller-manager/pkg/route/ingress	0.019s
FAIL

Reply:
Ah, you're right— the assignment `r.Annotations = existing.Annotations` overrides the annotations that were inherited from the Ingress with whatever annotations are set on the Route.  Why did you add the assignment?